File: DAA LAB/2nd program quick sort.cpp
#include <iostream>
using namespace std;

int count = 0;

void swap(int *a, int *b)
{
    int temp = *a;
    *a = *b;
    *b = temp;
}

int partition(int a[], int left, int right)
{
    int pivot = a[left], i = left + 1, j = right;
    while (1)
    {
        while (pivot >= a[i] && i <= right)
        {
            count++;
            i++;
        }
        while (pivot < a[j])
        {
            count++;
            j--;
        }
        if (i < j)
            swap(&a[i], &a[j]);
        else
        {
            swap(&a[left], &a[j]);
            return j;
        }
    }
}

void quick_sort(int a[], int left, int right)
{
    if (left < right)
    {
        int s = partition(a, left, right);
        quick_sort(a, left, s - 1);
        quick_sort(a, s + 1, right);
    }
}

int main()
{
    int n;
    cout << "Enter number of elements: ";
    cin >> n;
    int a[n], i;
    cout << "Enter " << n << " elements: ";
    for (i = 0; i < n; i++)
        cin >> a[i];
    quick_sort(a, 0, n - 1);
    cout << "Quick sorted array is:";
    for (i = 0; i < n; i++)
        cout << "\t" << a[i];
    cout << "\nNumber of times basic operation executed is: " << count;
    cout << "\n\nTime complexity table is:\n";
    cout << "size\tasc\trand\trand\n";
    int c1, c2, j;
    for (i = 1; i <= 4096; i *= 2)
    {
        int a[i], r[i], d[i];
        for (j = 0; j < i; j++)
        {
            a[j] = j + 1;
            r[j] = rand() % i;
            d[j] = i - j;
        }
        count = 0;
        quick_sort(a, 0, i - 1);
        c1 = count;
        count = 0;
        quick_sort(r, 0, i - 1);
        c2 = count;
        count = 0;
        quick_sort(d, 0, i - 1);
        cout << i << "\t" << c1 << "\t" << c2 << "\t" << count << endl;
    }
}



/*

Output of the code:

Enter number of elements: 5
Enter 5 elements: 1 54 1 6546 1
Quick sorted array is:  1       1       1       54      6546
Number of times basic operation executed is: 6

Time complexity table is:
size    asc     rand    rand
1       0       0       0
2       1       1       1
4       6       4       6
8       28      20      28
16      120     46      120
32      496     149     496
64      2016    361     2016
128     8128    967     8128
256     32640   1867    32640
512     130816  5224    130816
1024    523776  11487   523776
2048    2096128 24438   2096128
4096    8386560 61932   8386560



Enter number of elements: 12
Enter 12 elements:  1968 651 84 981 657 2 4 94 54 91 6 4
Quick sorted array is:  2       4       4       6       54      84      91      94      651     657     981    1968
Number of times basic operation executed is: 46

Time complexity table is:
size    asc     rand    rand
1       0       0       0
2       1       1       1
4       6       4       6
8       28      20      28
16      120     46      120
32      496     149     496
64      2016    361     2016
128     8128    967     8128
256     32640   1867    32640
512     130816  5224    130816
1024    523776  11487   523776
2048    2096128 24438   2096128
4096    8386560 61932   8386560


*/

File: DAA LAB/2nd program quick sort.java
import java.util.Random;
import java.util.Scanner;

public class QuickSort {

    static int count=0;

    public static void main(String[] args) {
        Scanner sc=new Scanner(System.in);
        System.out.print("Enter the number of elements: ");
        int n=sc.nextInt();
        int[] a=new int[n];
        System.out.print("Enter "+n+" elements: ");
        int i;
        for(i=0;i<n;i++)
            a[i]=sc.nextInt();
        quickSort(a,0,n-1);
        System.out.print("Quick sorted array is:");
        for(int var:a)
            System.out.print(" "+var);
        System.out.println("\nNumber of times basic operation executed is: "+count);
        int c1,c2,j;
        System.out.println("Time complexity table is");
        System.out.println("n\ta\tr\td");
        for(i=1;i<=4096;i*=2)
        {
            int[] asc=new int[i];
            int[] ran=new int[i];
            int[] des=new int[i];
            Random random=new Random(i);
            for(j=0;j<i;j++)
            {
                asc[j]=j+1;
                ran[j]=random.nextInt();
                des[j]=i-j;
            }
            count=0;
            quickSort(asc,0,i-1);
            c1=count;
            count=0;
            quickSort(ran,0,i-1);
            c2=count;
            count=0;
            quickSort(des,0,i-1);
            System.out.println(i+"\t"+c1+"\t"+c2+"\t"+count);
        }
    }

    static void quickSort(int[] a,int l,int r)
    {
        count++;
        if(l<r)
        {
            int s=partition(a,l,r);
            quickSort(a,l,s-1);
            quickSort(a,s+1,r);
        }
    }

    static int partition(int[] a,int l,int r)
    {
        int pivot=a[l],i=l+1,j=r;
        while(true)
        {
            count++;
            while(i<=r&&pivot>=a[i]) {
                i++;
            }
            while(pivot<a[j]) {
                j--;
            }
            if(i<j)
            {
                int temp=a[i];
                a[i]=a[j];
                a[j]=temp;
            }
            else{
                int temp=a[l];
                a[l]=a[j];
                a[j]=temp;
                return j;
            }
        }
    }
}

/*

Output of the code:

Enter the number of elements: 7
Enter 7 elements:  165 51 1 1 1 31 61 
Quick sorted array is: 1 1 1 31 51 61 165
Number of times basic operation executed is: 17
Time complexity table is
n	a	r	d
1	1	1	1
2	4	4	4
4	10	8	10
8	22	15	22
16	46	37	46
32	94	85	94
64	190	188	190
128	382	379	382
256	766	858	766
512	1534	1793	1534
1024	3070	3843	3070
2048	6142	8087	6142
4096	12286	17305	12286


Enter the number of elements: 3
Enter 3 elements: 1 4856 165 
Quick sorted array is: 1 165 4856
Number of times basic operation executed is: 7
Time complexity table is
n	a	r	d
1	1	1	1
2	4	4	4
4	10	8	10
8	22	15	22
16	46	37	46
32	94	85	94
64	190	188	190
128	382	379	382
256	766	858	766
512	1534	1793	1534
1024	3070	3843	3070
2048	6142	8087	6142
4096	12286	17305	12286


*/
